library/alloc/src/collections/btree/node.rs

```diff
@@ -59,31 +59,34 @@ struct LeafNode<K, V> {
 
     /// The number of keys and values this node stores.
     len: u16,
+}
+
+/// The underlying representation of leaf nodes and part of the representation of internal nodes.
+struct LeafNode<K, V> {
+    head: Head<K, V>,
 
     /// The arrays storing the actual data of the node. Only the first `len` elements of each
     /// array are initialized and valid.
     keys: [MaybeUninit<K>; CAPACITY],
     vals: [MaybeUninit<V>; CAPACITY],
 }
 
-impl<K, V> LeafNode<K, V> {
-    /// Initializes a new `LeafNode` in-place.
-    unsafe fn init(this: *mut Self) {
+impl<K, V> Head<K, V> {
+    /// Creates an initialized header.
+    fn new() -> Self {
         // As a general policy, we leave fields uninitialized if they can be, as this should
         // be both slightly faster and easier to track in Valgrind.
-        unsafe {
-            // parent_idx, keys, and vals are all MaybeUninit
-            ptr::addr_of_mut!((*this).parent).write(None);
-            ptr::addr_of_mut!((*this).len).write(0);
-        }
+        Head { parent: None, parent_idx: MaybeUninit::uninit(), len: 0 }
     }
+}
 
-    /// Creates a new boxed `LeafNode`. Unsafe because all nodes should really be hidden behind
-    /// `BoxedNode`, preventing accidental dropping of uninitialized keys and values.
-    unsafe fn new() -> Box<Self> {
+impl<K, V> LeafNode<K, V> {
+    /// Creates a new boxed `LeafNode`.
+    fn new() -> Box<Self> {
         unsafe {
-            let mut leaf = Box::new_uninit();
-            LeafNode::init(leaf.as_mut_ptr());
+            let mut leaf = Box::<Self>::new_uninit();
+            // We only need to initialize the header; the arrays are MaybeUninit.
+            (*leaf.as_mut_ptr()).head = Head::new();
```

Comment:
Should we still use a raw write here? There's no `Drop`, but I'm not sure if this still technically has implications about the former uninit `head` value.

Reply:
I don't really know how to do it non-raw. I'm happy when it compiles and when Miri says it the unit tests are okay. Although I haven't update or run the raw pointer tests lately, so I'm not sure Miri checks anything here.

Reply:
Oh, I think you mean the opposite of what I first understood. It isn't a raw write any more (because of #73987) and we probably should still do it.